src/mesh/voxel_cloud.rs

```diff
@@ -0,0 +1,499 @@
+use std::f32;
+use std::iter;
+
+use nalgebra::{Point3, Vector2, Vector3};
+
+use crate::convert::cast_usize;
+use crate::geometry;
+use crate::mesh::analysis::BoundingBox;
+use crate::mesh::primitive;
+use crate::mesh::tools;
+use crate::mesh::{Face, Mesh};
+use crate::plane::Plane;
+
+/// A public type for setting/getting a voxel state
+pub enum Voxel {
```

Comment:
how would you then write this?

```
match self.get_relative(&(voxel_coords + Vector3::new(0, 0, 1))) {
    // if there isn't or if the current voxel is on the
    // boundary of the voxel space block
    Voxel::Off | Voxel::OutOfBounds => {
        // add a horizontal rectangle
        plane_meshes.push(primitive::create_mesh_plane(
            Plane::from_origin_and_plane(
                // above the voxel center half way the height of the voxel
                &(voxel_center + upwards_voxel_direction),
                // align it properly
                &top_plane,
            ),
            // and set its size to match the dimensions
            // of the top side of a voxel
            top_bottom_voxel_dimensions,
        ));
    }
    // if there is a neighbor above the current voxel,
    // it means no boundary side of the voxel box should
    // be materialized
    _ => {}
}
```

Reply:
Out of bounds is not a voxel property. Coordinates can be out of bounds. Check if the coordinates are out of bounds OR if they are within bounds, check whether the voxel is off.

Reply:
...ok...

Reply:
If you really think this type is important, keep it *internally*, but never accept or return it from a public method.

Reply:
No, I'm going to change it to `Option<bool>` because it doesn't matter but I don't see why. Also why should I never return or accept it? We do it all over the place with normal strategies, errors, face types, so why not here?

Reply:
Ok, I'll try to explain, because this is important. This particular instance is not catastrophic, but it does add needless complexity that is not contained within the module.

This enum is an implementation detail, a sentinel value you use in internal logic to detect out of bounds. The caller of public methods of `VoxelCloud` doesn't need to see it at all. If all setters panic when you give them the `OutOfBounds` value and there never is a valid scenario for the setters to receive this value, simply don't make it possible in the typesystem for them to receive this value. For getters, there already is a precedence: in all rust's collection, the `get` method returns an option if the element is not in the collection or is out of bounds.

I think the enum was an attempt on your part to make symmetric the internal and externally visible functionality, but our goal should be to keep the program complexity down, so we want to hide what we can.

Reply:
> We do it all over the place with normal strategies, errors, face types, so why not here?

You misunderstand. This is not about this type being an enum. This is about this exact enum providing more freedom then strictly necessary.